libraries/freertos_plus/standard/utils/src/iot_pki_utils.c

```diff
@@ -0,0 +1,189 @@
+    uint8_t ucSigComponentLength = pxMbedSignature[ 3 ];
+
+    if( ( pxSignaturePKCS == NULL ) || ( pxMbedSignature == NULL ) )
+    {
+        xReturn = FAILURE;
+    }
+
+    if( xReturn == 0 )
+    {
+        /* The new signature will be 64 bytes long (32 bytes for R, 32 bytes for S).
+         * Zero this buffer out in case a component is shorter than 32 bytes. */
+        memset( pxSignaturePKCS, 0, 64 );
+
+        /********* R Component. *********/
+
+        /* R components are represented by mbedTLS as 33 bytes when the first bit is zero to avoid any sign confusion. */
+        if( ucSigComponentLength == 33 )
+        {
+            /* Chop off the leading zero.  The first 4 bytes were SEQUENCE, LENGTH, INTEGER, LENGTH, 0x00 padding.  */
+            memcpy( pxSignaturePKCS, &pxMbedSignature[ 5 ], 32 );
+            pxNextLength = pxMbedSignature + 5 /* SEQUENCE, LENGTH, INTEGER, LENGTH, leading zero */ + 32 /*(R) */ + 1 /*(S's integer tag) */;
+        }
+        else
+        {
+            /* The R component is 32 bytes or less.  Copy so that it is properly represented as a 32 byte value,
+             * leaving leading 0 pads at beginning if necessary. */
+            memcpy( &pxSignaturePKCS[ 32 - ucSigComponentLength ],         /* If the R component is less than 32 bytes, leave the leading zeros. */
+                    &pxMbedSignature[ 4 ],                                 /* SEQUENCE, LENGTH, INTEGER, LENGTH, (R component begins as the 5th byte) */
+                    ucSigComponentLength );
+            pxNextLength = pxMbedSignature + 4 + ucSigComponentLength + 1; /* Move the pointer to get rid of
+                                                                            * SEQUENCE, LENGTH, INTEGER, LENGTH, R Component, S integer tag. */
+        }
+
+        /********** S Component. ***********/
+
+        /* Now pxNextLength is pointing to the length of the S component. */
+        ucSigComponentLength = pxNextLength[ 0 ];
+
+        if( ucSigComponentLength == 33 )
+        {
+            memcpy( &pxSignaturePKCS[ 32 ],
+                    &pxNextLength[ 2 ], /*LENGTH (of S component), 0x00 padding, S component is 3rd byte - we want to skip the leading zero. */
+                    32 );
+        }
+        else
+        {
+            /* The S component is 32 bytes or less.  Copy so that it is properly represented as a 32 byte value,
+             * leaving leading 0 pads at beginning if necessary. */
+            memcpy( &pxSignaturePKCS[ 64 - ucSigComponentLength ],
+                    &pxNextLength[ 1 ],
+                    ucSigComponentLength );
+        }
+    }
+
+    return xReturn;
+}
+
+
+
+int PKI_pkcs11SignatureTombedTLSSignature( uint8_t * pucSig,
+                                           size_t * pxSigLen ) 
+{
+    int xResult = 0;
+    uint8_t * pucSigPtr;
+    uint8_t ucTemp[ 64 ] = { 0 }; /* A temporary buffer for the pre-formatted signature. */
```

Comment:
This routine is specific to ECDSA?

Reply:
yes.  I have added tiny comments above to reference this fact (and more complete explanation is in header).